File: sdk/src/main/java/eu/darken/porter/sdk/PorterProtocolWire.java
package eu.darken.porter.sdk;

import static eu.darken.porter.protocol.PorterProtocol.ATTACH_PACKAGE_NAME;
import static eu.darken.porter.protocol.PorterProtocol.ATTACH_PROTOCOL_VERSION;
import static eu.darken.porter.protocol.PorterProtocol.CAPABILITIES_NONE;
import static eu.darken.porter.protocol.PorterProtocol.PERMISSION_CONFIRMATION_ALLOWED;
import static eu.darken.porter.protocol.PorterProtocol.PERMISSION_CONFIRMATION_ONETIME;
import static eu.darken.porter.protocol.PorterProtocol.PERMISSION_RESULT_ALLOWED;
import static eu.darken.porter.protocol.PorterProtocol.REPLY_CAPABILITIES;
import static eu.darken.porter.protocol.PorterProtocol.REPLY_PERMISSION_GRANTED;
import static eu.darken.porter.protocol.PorterProtocol.REPLY_PROTOCOL_VERSION;
import static eu.darken.porter.protocol.PorterProtocol.REPLY_SERVER_SECONTEXT;
import static eu.darken.porter.protocol.PorterProtocol.REPLY_SERVER_UID;
import static eu.darken.porter.protocol.PorterProtocol.REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE;
import static eu.darken.porter.protocol.PorterProtocol.TRANSACTION_transactRemote;
import static eu.darken.porter.protocol.PorterProtocol.USER_SERVICE_TOKEN;

import android.os.Bundle;
import android.os.IBinder;
import android.os.Parcel;
import android.os.RemoteException;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import eu.darken.porter.protocol.PorterProtocol;
import eu.darken.porter.server.IPorterApplication;
import eu.darken.porter.server.IPorterService;
import eu.darken.porter.server.IPorterServiceConnection;

/** The Porter protocol as a client speaks it, over one server binder. */
final class PorterProtocolWire implements PorterWire {

    private final IPorterService service;
    private final Callbacks callbacks;

    /**
     * Held for the life of the connection: the server keeps only a proxy, so a stub that goes
     * unreachable here stops the pushes arriving.
     */
    private final IPorterApplication application = new IPorterApplication.Stub() {

        @Override
        public void dispatchRequestPermissionResult(int requestCode, Bundle data) {
            callbacks.onRequestPermissionResult(
                    requestCode, data.getBoolean(PERMISSION_RESULT_ALLOWED, false));
        }

        @Override
        public void dispatchPermissionStateChanged(Bundle state) {
            callbacks.onPermissionStateChanged(
                    state.getBoolean(REPLY_PERMISSION_GRANTED, false),
                    state.getBoolean(REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE, false));
        }
    };

    PorterProtocolWire(@NonNull IBinder binder, @NonNull Callbacks callbacks) {
        this.service = IPorterService.Stub.asInterface(binder);
        this.callbacks = callbacks;
    }

    @Nullable
    @Override
    public AttachReply attach(String packageName) throws RemoteException {
        Bundle args = new Bundle();
        args.putString(ATTACH_PACKAGE_NAME, packageName);
        args.putInt(ATTACH_PROTOCOL_VERSION, PorterProtocol.VERSION);

        Bundle reply = service.attach(application, args);
        if (reply == null) return null;
        return new AttachReply(
                reply.getInt(REPLY_SERVER_UID, -1),
                reply.getInt(REPLY_PROTOCOL_VERSION, 0),
                reply.getString(REPLY_SERVER_SECONTEXT),
                reply.getLong(REPLY_CAPABILITIES, CAPABILITIES_NONE),
                reply.getBoolean(REPLY_PERMISSION_GRANTED, false),
                reply.getBoolean(REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE, false),
                null);
    }

    @Override
    public void transactRemote(@NonNull Parcel data, @Nullable Parcel reply, int flags) {
        try {
            service.asBinder().transact(TRANSACTION_transactRemote, data, reply, flags);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public void forward(@NonNull IBinder target, int code, @NonNull Parcel data,
                        @Nullable Parcel reply, int flags) {
        // The token and the transaction are answered for by this one connection: a replacement
        // publishing between the two would have its server reject a parcel carrying the other
        // wire's token.
        Parcel newData = Parcel.obtain();
        try {
            newData.writeInterfaceToken(PorterProtocol.DESCRIPTOR);
            newData.writeStrongBinder(target);
            newData.writeInt(code);
            newData.writeInt(flags);
            newData.appendFrom(data, 0, data.dataSize());
            transactRemote(newData, reply, 0);
        } finally {
            newData.recycle();
        }
    }

    @Override
    public int getUid() {
        try {
            return service.getUid();
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public String getSELinuxContext() {
        try {
            return service.getSELinuxContext();
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public int checkPermission(String permission) {
        try {
            return service.checkPermission(permission);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public String getSystemProperty(String name, String defaultValue) {
        try {
            return service.getSystemProperty(name, defaultValue);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public void setSystemProperty(String name, String value) {
        try {
            service.setSystemProperty(name, value);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public void requestPermission(int requestCode) {
        try {
            service.requestPermission(requestCode);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public boolean checkSelfPermission() {
        try {
            return service.checkSelfPermission();
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public boolean shouldShowRequestPermissionRationale() {
        try {
            return service.shouldShowRequestPermissionRationale();
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public int addUserService(@NonNull UserServiceCallback conn,
                              @NonNull Porter.UserServiceArgs args, boolean noCreate) {
        IPorterServiceConnection adapter = adapterFor(conn);
        Bundle options = PorterUserServiceCodec.encodeUserService(args, noCreate);
        try {
            return service.addUserService(adapter, options);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public int removeUserService(@Nullable UserServiceCallback conn,
                                 @NonNull Porter.UserServiceArgs args, boolean remove) {
        IPorterServiceConnection adapter = adapterFor(conn);
        Bundle options = PorterUserServiceCodec.encodeUserServiceRemoval(args, remove);
        try {
            return service.removeUserService(adapter, options);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    /**
     * The stub this wire has registered for {@code callback}, created on first use, and null for the
     * removal that names no callback at all.
     *
     * <p>Lookup, creation and store happen under one lock on the callback, and the lock is released
     * before the call that carries the result: two threads binding one service register one stub.
     */
    @Nullable
    private static IPorterServiceConnection adapterFor(@Nullable UserServiceCallback callback) {
        if (callback == null) return null;
        synchronized (callback) {
            IBinder registered = callback.registeredBinder(PorterBackend.PORTER);
            if (registered != null) return IPorterServiceConnection.Stub.asInterface(registered);

            IPorterServiceConnection adapter = new IPorterServiceConnection.Stub() {

                @Override
                public void connected(IBinder binder) {
                    callback.connected(binder);
                }

                @Override
                public void died() {
                    callback.died();
                }
            };
            callback.rememberRegisteredBinder(PorterBackend.PORTER, adapter.asBinder());
            return adapter;
        }
    }

    @Override
    public void exit() {
        try {
            service.exit();
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public void attachUserService(@NonNull IBinder binder, @NonNull String token) {
        Bundle args = new Bundle();
        args.putString(USER_SERVICE_TOKEN, token);
        try {
            service.attachUserService(binder, args);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public void dispatchPermissionConfirmationResult(
            int requestUid, int requestPid, int requestCode, boolean allowed, boolean onetime) {
        Bundle data = new Bundle();
        data.putBoolean(PERMISSION_CONFIRMATION_ALLOWED, allowed);
        data.putBoolean(PERMISSION_CONFIRMATION_ONETIME, onetime);
        try {
            service.dispatchPermissionConfirmationResult(requestUid, requestPid, requestCode, data);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public int getFlagsForUid(int uid, int mask) {
        try {
            return service.getFlagsForUid(uid, mask);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    @Override
    public void updateFlagsForUid(int uid, int mask, int value) {
        try {
            service.updateFlagsForUid(uid, mask, value);
        } catch (RemoteException e) {
            throw rethrowAsRuntimeException(e);
        }
    }

    private static RuntimeException rethrowAsRuntimeException(RemoteException e) {
        return new RuntimeException(e);
    }
}

File: sdk/src/main/java/eu/darken/porter/sdk/PorterProtocolWire.kt
package eu.darken.porter.sdk

import android.os.Bundle
import android.os.IBinder
import android.os.Parcel
import android.os.RemoteException
import eu.darken.porter.protocol.PorterProtocol
import eu.darken.porter.protocol.PorterProtocol.ATTACH_PACKAGE_NAME
import eu.darken.porter.protocol.PorterProtocol.ATTACH_PROTOCOL_VERSION
import eu.darken.porter.protocol.PorterProtocol.CAPABILITIES_NONE
import eu.darken.porter.protocol.PorterProtocol.PERMISSION_RESULT_ALLOWED
import eu.darken.porter.protocol.PorterProtocol.REPLY_CAPABILITIES
import eu.darken.porter.protocol.PorterProtocol.REPLY_MIN_PROTOCOL_VERSION
import eu.darken.porter.protocol.PorterProtocol.REPLY_PERMISSION_GRANTED
import eu.darken.porter.protocol.PorterProtocol.REPLY_PROTOCOL_VERSION
import eu.darken.porter.protocol.PorterProtocol.REPLY_SERVER_SECONTEXT
import eu.darken.porter.protocol.PorterProtocol.REPLY_SERVER_UID
import eu.darken.porter.protocol.PorterProtocol.REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE
import eu.darken.porter.protocol.PorterProtocol.REPLY_UNSUPPORTED
import eu.darken.porter.protocol.PorterProtocol.TRANSACTION_transactRemote
import eu.darken.porter.server.IPorterApplication
import eu.darken.porter.server.IPorterService
import eu.darken.porter.server.IPorterServiceConnection

/** The Porter protocol as a client speaks it, over one server binder. */
internal class PorterProtocolWire(
    binder: IBinder,
    private val callbacks: PorterWire.Callbacks,
) : PorterWire {

    private val service: IPorterService = IPorterService.Stub.asInterface(binder)

    /**
     * Held for the life of the connection: the server keeps only a proxy, so a stub that goes
     * unreachable here stops the pushes arriving.
     */
    private val application: IPorterApplication = object : IPorterApplication.Stub() {

        override fun dispatchRequestPermissionResult(requestCode: Int, data: Bundle) {
            callbacks.onRequestPermissionResult(
                requestCode, data.getBoolean(PERMISSION_RESULT_ALLOWED, false),
            )
        }

        override fun dispatchPermissionStateChanged(state: Bundle) {
            callbacks.onPermissionStateChanged(
                state.getBoolean(REPLY_PERMISSION_GRANTED, false),
                state.getBoolean(REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE, false),
            )
        }
    }

    @Throws(RemoteException::class)
    override fun attach(packageName: String): PorterWire.AttachReply? {
        val args = Bundle().apply {
            putString(ATTACH_PACKAGE_NAME, packageName)
            putInt(ATTACH_PROTOCOL_VERSION, PorterProtocol.VERSION)
        }

        val reply = service.attach(application, args) ?: return null
        return PorterWire.AttachReply(
            serverUid = reply.getInt(REPLY_SERVER_UID, -1),
            protocolVersion = reply.getInt(REPLY_PROTOCOL_VERSION, 0),
            seLinuxContext = reply.getString(REPLY_SERVER_SECONTEXT),
            capabilities = reply.getLong(REPLY_CAPABILITIES, CAPABILITIES_NONE),
            permissionGranted = reply.getBoolean(REPLY_PERMISSION_GRANTED, false),
            shouldShowRequestPermissionRationale =
                reply.getBoolean(REPLY_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE, false),
            patchVersion = null,
            minProtocolVersion = reply.getInt(REPLY_MIN_PROTOCOL_VERSION, 0),
            unsupported = reply.getBoolean(REPLY_UNSUPPORTED, false),
        )
    }

    override fun incompatibility(reply: PorterWire.AttachReply?): PorterIncompatibility? {
        if (reply == null) return PorterIncompatibility(serverVersion = 0, serverMinVersion = 0)
        val mismatch = PorterIncompatibility(reply.protocolVersion, reply.minProtocolVersion)
        // A server that refused says so; one that did not is still held to this side's floor, and
        // one that reports no version at all cannot be told apart from one below it.
        val refused = reply.unsupported || reply.protocolVersion < MIN_SERVER_VERSION
        return if (refused || mismatch.serverTooOld || mismatch.clientTooOld) mismatch else null
    }

    override fun transactRemote(data: Parcel, reply: Parcel?, flags: Int) {
        remote { service.asBinder().transact(TRANSACTION_transactRemote, data, reply, flags) }
    }

    override fun forward(target: IBinder, code: Int, data: Parcel, reply: Parcel?, flags: Int) {
        // The token and the transaction are answered for by this one connection: a replacement
        // publishing between the two would have its server reject a parcel carrying the other
        // wire's token.
        val newData = Parcel.obtain()
        try {
            newData.writeInterfaceToken(PorterProtocol.DESCRIPTOR)
            newData.writeStrongBinder(target)
            newData.writeInt(code)
            newData.writeInt(flags)
            newData.appendFrom(data, 0, data.dataSize())
            transactRemote(newData, reply, 0)
        } finally {
            newData.recycle()
        }
    }

    override fun getUid(): Int = remote { service.getUid() }

    override fun getSELinuxContext(): String? = remote { service.getSELinuxContext() }

    override fun checkPermission(permission: String): Int = remote { service.checkPermission(permission) }

    override fun getSystemProperty(name: String, defaultValue: String?): String? =
        remote { service.getSystemProperty(name, defaultValue) }

    override fun setSystemProperty(name: String, value: String) {
        remote { service.setSystemProperty(name, value) }
    }

    override fun requestPermission(requestCode: Int) {
        remote { service.requestPermission(requestCode) }
    }

    override fun checkSelfPermission(): Boolean = remote { service.checkSelfPermission() }

    override fun shouldShowRequestPermissionRationale(): Boolean =
        remote { service.shouldShowRequestPermissionRationale() }

    override fun addUserService(conn: UserServiceCallback, args: UserServiceArgs, noCreate: Boolean): Int {
        val adapter = adapterFor(conn)
        val options = PorterUserServiceCodec.encodeUserService(args, noCreate)
        return remote { service.addUserService(adapter, options) }
    }

    override fun removeUserService(conn: UserServiceCallback?, args: UserServiceArgs, remove: Boolean): Int {
        val adapter = adapterFor(conn)
        val options = PorterUserServiceCodec.encodeUserServiceRemoval(args, remove)
        return remote { service.removeUserService(adapter, options) }
    }

    private inline fun <T> remote(call: () -> T): T = try {
        call()
    } catch (e: RemoteException) {
        throw PorterRemoteException(e)
    }

    internal companion object {

        /**
         * The oldest server this SDK still speaks to. Raised only when the SDK stops speaking an
         * older generation; a server above [PorterProtocol.VERSION] is fine as long as its own
         * floor admits this client.
         */
        const val MIN_SERVER_VERSION: Int = 4

        /**
         * The stub this wire has registered for [callback], created on first use, and null for the
         * removal that names no callback at all.
         *
         * Lookup, creation and store happen under one lock on the callback, and the lock is released
         * before the call that carries the result: two threads binding one service register one stub.
         */
        fun adapterFor(callback: UserServiceCallback?): IPorterServiceConnection? {
            if (callback == null) return null
            synchronized(callback) {
                val registered = callback.registeredBinder(PorterBackend.PORTER)
                if (registered != null) return IPorterServiceConnection.Stub.asInterface(registered)

                val adapter = object : IPorterServiceConnection.Stub() {

                    override fun connected(binder: IBinder) {
                        callback.connected(binder)
                    }

                    override fun died() {
                        callback.died()
                    }
                }
                callback.rememberRegisteredBinder(PorterBackend.PORTER, adapter.asBinder())
                return adapter
            }
        }
    }
}
